utils/writeable/src/impls.rs

```diff
@@ -8,84 +8,103 @@ use core::convert::TryFrom;
 
-impl Writeable for u16 {
-    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
-        let mut buf = [b'0'; 5];
-        let mut n = *self;
-        let mut i = 5usize;
-        while n != 0 {
-            i -= 1;
-            buf[i] = b'0' + u8::try_from(n % 10).expect("<10");
-            n /= 10;
-        }
-        if i == 5 {
-            debug_assert_eq!(*self, 0);
-            i = 4;
+                // If the bounds aren't tight (e.g. 87 ∈ [64, 127] ⟹ len ∈ [2,3]), compare
+                // to 10ʰ⁻¹ (100). This shouldn't happen too often as there are more powers
+                // of 2 than 10 (it happens for 14% of u32s).
+                if low == high || *self < (10 as $u).pow(low) {
+                    LengthHint::exact(low as usize)
+                } else {
+                    LengthHint::exact(high as usize)
+                }
+            }
         }
-        let s = unsafe { str::from_utf8_unchecked(&buf[i..]) };
-        sink.write_str(s)
-    }
 
-    fn write_len(&self) -> LengthHint {
-        if *self < 10 {
-            LengthHint::exact(1)
-        } else if *self < 100 {
-            LengthHint::exact(2)
-        } else if *self < 1000 {
-            LengthHint::exact(3)
-        } else if *self < 10000 {
-            LengthHint::exact(4)
-        } else {
-            LengthHint::exact(5)
+        impl Writeable for $i {
+            fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+                if self.is_negative() {
+                    sink.write_str("-")?;
+                }
+                self.unsigned_abs().write_to(sink)
+            }
+
+            fn write_len(&self) -> LengthHint {
+                LengthHint::exact(if self.is_negative() { 1 } else { 0 })
+                    + self.unsigned_abs().write_len()
+            }
         }
-    }
+    };
 }
 
+impl_write_num!(u8, i8);
+impl_write_num!(u16, i16);
+impl_write_num!(u32, i32);
+impl_write_num!(u64, i64);
+impl_write_num!(u128, i128);
+
 #[test]
-fn test_u8() {
+fn test_ints() {
     use crate::assert_writeable_eq;
-    assert_writeable_eq!(&0u8, "0",);
-    assert_writeable_eq!(&1u8, "1",);
-    assert_writeable_eq!(&10u8, "10",);
-    assert_writeable_eq!(&99u8, "99",);
-    assert_writeable_eq!(&255u8, "255",);
+    assert_writeable_eq!(&0u128, "0");
+    assert_writeable_eq!(&0i128, "0");
+    assert_writeable_eq!(&-0i128, "0");
+    assert_writeable_eq!(&1u128, "1");
+    assert_writeable_eq!(&1i128, "1");
+    assert_writeable_eq!(&-1i128, "-1");
+    assert_writeable_eq!(&10u128, "10");
+    assert_writeable_eq!(&10i128, "10");
+    assert_writeable_eq!(&-10i128, "-10");
+    assert_writeable_eq!(&99u128, "99");
+    assert_writeable_eq!(&99i128, "99");
+    assert_writeable_eq!(&-99i128, "-99");
+    assert_writeable_eq!(&u128::MAX, u128::MAX.to_string());
+    assert_writeable_eq!(&i128::MAX, i128::MAX.to_string());
+    assert_writeable_eq!(&i128::MIN, i128::MIN.to_string());
 }
```

Comment:
Nit: Add more tests.  Suggestions:

- More tests around power-of-10 boundaries (99, 100, 999, 1000, ...)
- Randomly generated integers of each size (u8, u16, u32, u64, u128) that check the Writeable impl against the ToString impl

This can be done in a follow-up PR if you need to merge this one for a dependency.